Drivers/Src/PWM.c

```diff
@@ -0,0 +1,155 @@
+}
+
+QueueHandle_t* BSP_PWM_Get_Queue(TIM_HandleTypeDef* timHandle) {
+    switch((uint32_t)timHandle->Instance) {
+        case (uint32_t)TIM1:
+            return &pwm1_send_queue;
+        case (uint32_t)TIM9:
+            return &pwm2_send_queue;
+        //add more timers later maybe
+    }
+    return 0;
+}
+
+HAL_StatusTypeDef BSP_PWM_TIM_Init(TIM_HandleTypeDef* timHandle) {
+    HAL_StatusTypeDef stat = HAL_OK;
+    
+    if(__HAL_RCC_GPIOA_IS_CLK_DISABLED())
+         __HAL_RCC_GPIOA_CLK_ENABLE();
+    
+    if (timHandle->Instance == TIM1 && pwm1_send_queue == NULL) {// assign queue for specific timer
+        pwm1_send_queue = xQueueCreateStatic(PWM_SEND_QUEUE_SIZE, PWM_INFO_SIZE,
+        pwm1_send_queue_storage, &pwm1_send_queue_buffer);
+    
+        if (pwm1_send_queue == NULL) 
+            return HAL_ERROR; 
+    }
+    else if (timHandle->Instance == TIM9 && pwm2_send_queue == NULL) {
+        
+        pwm2_send_queue = xQueueCreateStatic(PWM_SEND_QUEUE_SIZE, PWM_INFO_SIZE,
+        pwm2_send_queue_storage, &pwm2_send_queue_buffer); 
+
+        if (pwm2_send_queue == NULL) 
+            return HAL_ERROR; 
+    }
+
+    stat = HAL_TIM_PWM_Init(timHandle);
+    if(stat == HAL_ERROR) return stat;
+
+    return stat;
+}
+
+HAL_StatusTypeDef BSP_PWM_Channel_Init(TIM_HandleTypeDef* timHandle, uint8_t channel) {    
+    HAL_StatusTypeDef stat = HAL_OK;
+    
+    sConfigOC.OCMode = TIM_OCMODE_PWM1;  
+    sConfigOC.Pulse = 5000;
+    sConfigOC.OCPolarity = TIM_OCPOLARITY_HIGH;
+    sConfigOC.OCFastMode = TIM_OCFAST_DISABLE;
+
+    stat = HAL_TIM_PWM_ConfigChannel(timHandle, &sConfigOC, channel);
+    if(stat == HAL_ERROR) return stat;
+
+    stat = HAL_TIM_PWM_Start_IT(timHandle, channel); 
+    return stat;
+}
+
+HAL_StatusTypeDef BSP_PWM_Set(TIM_HandleTypeDef* timHandle, uint8_t channel, uint8_t dutyCycle, uint64_t speed) {
+    if(dutyCycle <= 100 && dutyCycle >= 0) {
+
+        QueueHandle_t* tx_Queue = BSP_PWM_Get_Queue(timHandle);
+    
+        PWM_Info pwmSend = { // for storing PWM info into queue
+            .timHandle = timHandle,
+            .channel = channel,
+            .dutyCycle = dutyCycle,
+            .speed = speed
+        };
+
+        if(!xQueueSend(*tx_Queue, &pwmSend, 0)) {
+            return HAL_ERROR;
+        }
+            
+        return HAL_OK;
+    }
+    return HAL_ERROR;
+}
+
+void HAL_TIM_PWM_PeriodElapsedCallback(TIM_HandleTypeDef *timHandle) {
+
+    // HAL_GPIO_TogglePin(GPIOA, GPIO_PIN_5); // testing interrupt callback
+    QueueHandle_t* tx_Queue = BSP_PWM_Get_Queue(timHandle);
+ 
+
+    if(!xQueueIsQueueEmptyFromISR(*tx_Queue)) {
+        PWM_Info pwmReceive; 
+        xQueueReceiveFromISR(*tx_Queue, &pwmReceive, &higherPriorityTaskWoken);
+        HAL_TIM_PWM_Stop_IT(pwmReceive.timHandle, pwmReceive.channel);
+        // pwmReceive.timHandle->Init.Period = pwmReceive.speed;
+        sConfigOC.Pulse = pwmReceive.dutyCycle*(pwmReceive.timHandle->Init.Period)/100;
```

Comment:
make sure the multiplication happens first here just by surrounding it in parentheses, so that there is no unintended precision loss from dividing first.
```suggestion
        sConfigOC.Pulse = (pwmReceive.dutyCycle*pwmReceive.timHandle->Init.Period)/100;
```